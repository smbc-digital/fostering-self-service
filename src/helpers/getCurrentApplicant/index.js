import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant