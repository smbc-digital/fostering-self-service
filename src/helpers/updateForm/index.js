import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}