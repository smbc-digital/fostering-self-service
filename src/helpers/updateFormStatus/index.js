export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}