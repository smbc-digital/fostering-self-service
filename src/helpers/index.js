import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}