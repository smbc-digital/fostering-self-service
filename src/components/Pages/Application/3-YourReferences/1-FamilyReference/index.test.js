import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import FamilyReference from './index'
import { TaskStatus } from 'helpers'
import { getPageRoute } from 'helpers'

describe('FamilyReference', () => {

	const onChangeStatusMock = jest.fn()
	const onChangeApplicantMock = jest.fn()

	beforeEach(() => {
		jest.resetAllMocks()
		useContextMock.mockReturnValue({
			statuses: {
				referencesStatus: TaskStatus.Completed
			},
			onChangeStatus: onChangeStatusMock,
			onChangeApplicant: onChangeApplicantMock,
			firstApplicant: {
				firstName: {
                    value: 'test',
                    isValid: true
                },
                lastName: {
                    value: 'test',
                    isValid: true
                }
			},
			secondApplicant: null,
			familyReference: {
				firstName: {
                    value: 'refname',
                    isValid: true
                },
                lastName: {
                    value: 'reflastname',
                    isValid: true
				},
				relationshipToYou: {
					value: 'relationship',
					isValid: true
				},
				emailAddress: {
					value: 'email',
					isValid: true
				},
				phoneNumber: {
					value: 'phone',
					isValid: true
				},
				address: {
					value: {
						placeRef: '123'
					},
					isValid: true
				}
			}
		})
	})

	it('should push to next page on submit', () => {
		// Arrange
		const history = {
			push: jest.fn()
		}

		const wrapper = mount(<FamilyReference history={history}/>)

		// Act
		wrapper.find('Button').at(0).simulate('submit')

		// Assert
		expect(history.push).toHaveBeenCalledWith(getPageRoute(24))
	})

	it('should call onChangeApplicant', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        // Act
        const wrapper = mount(<FamilyReference history={history} />)
        wrapper.find('input').first().simulate('change')

        // Assert
        expect(onChangeApplicantMock).toHaveBeenCalled()
	})
	
	describe('snapshot', () => {
        
        it('should render correctly for single applicant', () => {
            const tree = renderer
                .create(<FamilyReference history={{}} />)
                .toJSON()

            expect(tree).toMatchSnapshot()
		})
		
		it('should render correctly for joint applicant', () => {
			useContextMock.mockReturnValue({
				statuses: {
				referencesStatus: TaskStatus.Completed
				},
				onChangeStatus: onChangeStatusMock,
				onChangeApplicant: onChangeApplicantMock,
				firstApplicant: {
					firstName: {
						value: 'test',
						isValid: true
					},
					lastName: {
						value: 'test',
						isValid: true
					}
				},
				secondApplicant: {
					firstName: {
						value: 'test',
						isValid: true
					},
					lastName: {
						value: 'test',
						isValid: true
					}
				},
				familyReference: {
					firstName: {
						value: 'refname',
						isValid: true
					},
					lastName: {
						value: 'reflastname',
						isValid: true
					},
					relationshipToYou: {
						value: 'relationship',
						isValid: true
					},
					emailAddress: {
						value: 'email',
						isValid: true
					},
					phoneNumber: {
						value: 'phone',
						isValid: true
					},
					address: {
						value: {
							placeRef: '123'
						},
						isValid: true
					}
				}
			})

			const tree = renderer
			.create(<FamilyReference history={{}} />)
			.toJSON()

		expect(tree).toMatchSnapshot()
		})
    })
})