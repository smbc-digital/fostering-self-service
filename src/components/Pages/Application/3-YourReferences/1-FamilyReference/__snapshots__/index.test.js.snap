// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FamilyReference snapshot should render correctly for joint applicant 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Personal references
  </h2>,
  <h3>
    Tell us about the relative who can give a personal reference for you and your partner
  </h3>,
  <div>
    <section
      className="alert-information-form"
    >
      <span
        className="fa fa-info-circle"
      />
      <p
        dangerouslySetInnerHTML={
          Object {
            "__html": "Your referee must be related to you or your partner and be able to comment on your home life and your contact with children.",
          }
        }
      />
    </section>
  </div>,
  <form
    onSubmit={[Function]}
  >
    <label
      className=""
      htmlFor="firstName"
    >
      First name
    </label>
    <input
      aria-label="First name"
      aria-live="off"
      className=""
      id="firstName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="firstName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="refname"
    />
    <label
      className=""
      htmlFor="lastName"
    >
      Last name
    </label>
    <input
      aria-label="Last name"
      aria-live="off"
      className=""
      id="lastName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="lastName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="reflastname"
    />
    <label
      className=""
      htmlFor="relationshipToYou"
    >
      Relationship to you
    </label>
    <input
      aria-label="Relationship to you"
      aria-live="off"
      className=""
      id="relationshipToYou"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="relationshipToYou"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="relationship"
    />
    <label
      className=""
      htmlFor="emailAddress"
    >
      Email address
    </label>
    <input
      aria-label="Email address"
      aria-live="off"
      className=""
      id="emailAddress"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="emailAddress"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="email"
      value="email"
    />
    <label
      className=""
      htmlFor="phoneNumber"
    >
      Phone number
    </label>
    <input
      aria-label="Phone number"
      aria-live="off"
      className=""
      id="phoneNumber"
      inputMode={undefined}
      maxLength={11}
      min={undefined}
      name="phoneNumber"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="tel"
      value="phone"
    />
    <label
      className=""
      htmlFor="postcode"
    >
      Enter the postcode
    </label>
    <input
      aria-label="Enter the postcode"
      aria-live="off"
      className=" address-picker-input"
      id="postcode"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="postcode"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value=""
    />
    <button
      className="button-secondary"
      disabled={true}
      onClick={[Function]}
    >
      Find the address
    </button>
    <p>
      <a
        className="anchor-as-block"
        href="#"
        onClick={[Function]}
      >
        Enter the address manually
      </a>
    </p>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
  </form>,
]
`;

exports[`FamilyReference snapshot should render correctly for single applicant 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Personal references
  </h2>,
  <h3>
    Tell us about the relative who can give a personal reference for you
  </h3>,
  <div>
    <section
      className="alert-information-form"
    >
      <span
        className="fa fa-info-circle"
      />
      <p
        dangerouslySetInnerHTML={
          Object {
            "__html": "Your referee must be related to you and be able to comment on your home life and your contact with children.",
          }
        }
      />
    </section>
  </div>,
  <form
    onSubmit={[Function]}
  >
    <label
      className=""
      htmlFor="firstName"
    >
      First name
    </label>
    <input
      aria-label="First name"
      aria-live="off"
      className=""
      id="firstName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="firstName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="refname"
    />
    <label
      className=""
      htmlFor="lastName"
    >
      Last name
    </label>
    <input
      aria-label="Last name"
      aria-live="off"
      className=""
      id="lastName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="lastName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="reflastname"
    />
    <label
      className=""
      htmlFor="relationshipToYou"
    >
      Relationship to you
    </label>
    <input
      aria-label="Relationship to you"
      aria-live="off"
      className=""
      id="relationshipToYou"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="relationshipToYou"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="relationship"
    />
    <label
      className=""
      htmlFor="emailAddress"
    >
      Email address
    </label>
    <input
      aria-label="Email address"
      aria-live="off"
      className=""
      id="emailAddress"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="emailAddress"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="email"
      value="email"
    />
    <label
      className=""
      htmlFor="phoneNumber"
    >
      Phone number
    </label>
    <input
      aria-label="Phone number"
      aria-live="off"
      className=""
      id="phoneNumber"
      inputMode={undefined}
      maxLength={11}
      min={undefined}
      name="phoneNumber"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="tel"
      value="phone"
    />
    <label
      className=""
      htmlFor="postcode"
    >
      Enter the postcode
    </label>
    <input
      aria-label="Enter the postcode"
      aria-live="off"
      className=" address-picker-input"
      id="postcode"
      inputMode={undefined}
      maxLength={undefined}
      min={undefined}
      name="postcode"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value=""
    />
    <button
      className="button-secondary"
      disabled={true}
      onClick={[Function]}
    >
      Find the address
    </button>
    <p>
      <a
        className="anchor-as-block"
        href="#"
        onClick={[Function]}
      >
        Enter the address manually
      </a>
    </p>
    <button
      className="button-primary"
      disabled={true}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
  </form>,
]
`;
